pkg/api/volume/formatter.go

```diff
@@ -0,0 +1,87 @@
+package volume
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/apiserver/pkg/types"
+	v1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	"github.com/sirupsen/logrus"
+
+	"github.com/harvester/harvester/pkg/api/vm"
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	"github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1"
+)
+
+const (
+	actionExport = "export"
+)
+
+func Formatter(request *types.APIRequest, resource *types.RawResource) {
+	resource.Actions = make(map[string]string, 1)
+	resource.AddAction(request, actionExport)
+}
+
+type ExportActionHandler struct {
+	images   v1beta1.VirtualMachineImageClient
+	pvcCache v1.PersistentVolumeClaimCache
+}
+
+func (h ExportActionHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if err := h.do(rw, req); err != nil {
+		status := http.StatusInternalServerError
+		if e, ok := err.(*apierror.APIError); ok {
+			status = e.Code.Status
+		}
+		rw.WriteHeader(status)
+		_, _ = rw.Write([]byte(err.Error()))
+		return
+	}
+	rw.WriteHeader(http.StatusOK)
+}
+
+func (h ExportActionHandler) do(rw http.ResponseWriter, r *http.Request) error {
+	vars := mux.Vars(r)
+	action := vars["action"]
+	namespace := vars["namespace"]
+	pvcName := vars["name"]
+
+	switch action {
+	case actionExport:
+		var input vm.ExportVolumeInput
+		if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
+			return apierror.NewAPIError(validation.InvalidBodyContent, "Failed to decode request body: %v "+err.Error())
+		}
+		if input.DisplayName == "" {
+			return apierror.NewAPIError(validation.InvalidBodyContent, "Parameter `displayName` are required")
+		}
+		return h.exportVolume(r.Context(), namespace, input.DisplayName, pvcName)
+	default:
+		return apierror.NewAPIError(validation.InvalidAction, "Unsupported action")
+	}
+}
+
+func (h ExportActionHandler) exportVolume(ctx context.Context, namespace, displayName, pvcName string) error {
+	pvc, err := h.pvcCache.Get(namespace, pvcName)
+	if err != nil {
+		return fmt.Errorf("failed to get pvc %s/%s, error: %s", pvcName, namespace, err.Error())
+	}
+	vmi := &harvesterv1.VirtualMachineImage{
+		Spec: harvesterv1.VirtualMachineImageSpec{
+			DisplayName: displayName,
+			SourceType:  harvesterv1.VirtualMachineImageSourceTypeExportVolume,
+			VolumeName:  pvc.Spec.VolumeName,
```

Comment:
I think we need to use PVCName for this image field. `pvc.Spec.VolumeName(PV)` is not something we expose to the users. We can fetch the PersistentVolumeName in the controller and set it to backing images.

Reply:
@gitlawr updated.